Mail.1
.\"	$OpenBSD: mail.1,v 1.5 1994/06/29 05:09:32 deraadt Exp $
.\" Copyright (c) 1980, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	from: @(#)mail.1	8.2 (Berkeley) 12/30/93
.\"
.Dd December 30, 1993
.Dt MAIL 1
.Os BSD 4
.Sh NAME
.Nm mail
.Nd send and receive mail
.Sh SYNOPSIS
.Nm mail
.Op Fl iInv
.Op Fl s Ar subject
.Op Fl c Ar cc-addr
.Op Fl b Ar bcc-addr
.Ar to-addr...
.Nm mail
.Op Fl iInNv
.Fl f
.Op Ar name
.Nm mail
.Op Fl iInNv
.Op Fl u Ar user
.Sh INTRODUCTION
.Nm Mail
is an intelligent mail processing system, which has
a command syntax reminiscent of
.Xr \&ed 1
with lines replaced by messages.
.Pp
.Bl -tag -width flag
.It Fl v
Verbose mode.
The details of
delivery are displayed on the user's terminal.
.It Fl i
Ignore tty interrupt signals.
This is
particularly useful when using
.Nm mail
on noisy phone lines.
.It Fl I
Forces mail to run in interactive mode even when
input isn't a terminal.
In particular, the
.Sq Ic \&~
special
character when sending mail is only active in interactive mode.
.It Fl n
Inhibits reading
.Pa /etc/mail.rc
upon startup.
.It Fl N
Inhibits the initial display of message headers
when reading mail or editing a mail folder.
.It Fl s
Specify subject on command line
(only the first argument after the
.Fl s
flag is used as a subject; be careful to quote subjects
containing spaces.)
.It Fl c
Send carbon copies to
.Ar list
of users.
.It Fl b
Send blind carbon copies to
.Ar list .
List should be a comma-separated list of names.
.It Fl f
Read in the contents of your
.Ar mbox
(or the specified file)
for processing; when you
.Ar quit  ,
.Nm mail
writes undeleted messages back to this file.
.It Fl u
Is equivalent to:
.Pp
.Dl mail -f /var/spool/mail/user
.El
.Ss Sending mail
To send a message to one or more people,
.Nm mail
can be invoked with arguments which are the names of people to
whom the mail will be sent.
You are then expected to type in
your message, followed
by an
.Sq Li control\-D
at the beginning of a line.
The section below
.Ar Replying to or originating mail ,
describes some features of
.Nm mail
available to help you compose your letter.
.Pp
.Ss Reading mail
In normal usage
.Nm mail
is given no arguments and checks your mail out of the
post office, then
prints out a one line header of each message found.
The current message is initially the first message (numbered 1)
and can be printed using the
.Ic print
command (which can be abbreviated
.Ql Ic p ) .
You can move among the messages much as you move between lines in
.Xr \&ed 1 ,
with the commands
.Ql Ic \&+
and
.Ql Ic \&\-
moving backwards and forwards, and
simple numbers.
.Pp
.Ss Disposing of mail.
After examining a message you can
.Ic delete
.Ql Ic d )
the message or
.Ic reply
.Ql Ic r )
to it.
Deletion causes the
.Nm mail
program to forget about the message.
This is not irreversible; the message can be
.Ic undeleted
.Ql Ic u )
by giving its number, or the
.Nm mail
session can be aborted by giving the
.Ic exit
.Ql Ic x )
command.
Deleted messages will, however, usually disappear never to be seen again.
.Pp
.Ss Specifying messages
Commands such as
.Ic print
and
.Ic delete
can be given a list of message numbers as arguments to apply
to a number of messages at once.
Thus
.Dq Li delete 1 2
deletes messages 1 and 2, while
.Dq Li delete 1\-5
deletes messages 1 through 5.
The special name
.Ql Li \&*
addresses all messages, and
.Ql Li \&$
addresses
the last message; thus the command
.Ic top
which prints the first few lines of a message could be used in
.Dq Li top \&*
to print the first few lines of all messages.
.Pp
.Ss Replying to or originating mail.
You can use the
.Ic reply
command to
set up a response to a message, sending it back to the
person who it was from.
Text you then type in, up to an end-of-file,
defines the contents of the message.
While you are composing a message,
.Nm mail
treats lines beginning with the character
.Ql Ic \&~
specially.
For instance, typing
.Ql Ic \&~m
(alone on a line) will place a copy
of the current message into the response right shifting it by a tabstop
(see
.Em indentprefix
variable, below).
Other escapes will set up subject fields, add and delete recipients
to the message and allow you to escape to an editor to revise the
message or to a shell to run some commands.
(These options
are given in the summary below.)
.Pp
.Ss Ending a mail processing session.
You can end a
.Nm mail
session with the
.Ic quit
.Ql Ic q )
command.
Messages which have been examined go to your
.Ar mbox
file unless they have been deleted in which case they are discarded.
Unexamined messages go back to the post office.
(See the
.Fl f
option above).
.Pp
.Ss Personal and systemwide distribution lists.
It is also possible to create a personal distribution lists so that,
for instance, you can send mail to
.Dq Li cohorts
and have it go
to a group of people.
Such lists can be defined by placing a line like
.Pp
.Dl alias cohorts bill ozalp jkf mark kridle@ucbcory
.Pp
in the file
.Pa \&.mailrc
in your home directory.
The current list of such aliases can be displayed with the
.Ic alias
command in
.Nm mail  .
System wide distribution lists can be created by editing
.Pa /etc/aliases ,
see
.Xr aliases  5
and
.Xr sendmail  8  ;
these are kept in a different syntax.
In mail you send, personal aliases will be expanded in mail sent
to others so that they will be able to
.Ic reply
to the recipients.
System wide
.Ic aliases
are not expanded when the mail is sent,
but any reply returned to the machine will have the system wide
alias expanded as all mail goes through
.Xr sendmail  .
.Pp
.Ss Network mail (ARPA, UUCP, Berknet)
See
.Xr mailaddr 7
for a description of network addresses.
.Pp
.Nm Mail
has a number of options which can be set in the
.Pa .mailrc
file to alter its behavior; thus
.Dq Li set askcc
enables the
.Ar askcc
feature.
(These options are summarized below.)
.Sh SUMMARY
(Adapted from the `Mail Reference Manual')
.Pp
Each command is typed on a line by itself, and may take arguments
following the command word.
The command need not be typed in its
entirety \- the first command which matches the typed prefix is used.
For commands which take message lists as arguments, if no message
list is given, then the next message forward which satisfies the
command's requirements is used.
If there are no messages forward of
the current message, the search proceeds backwards, and if there are no
good messages at all,
.Nm mail
types
.Dq Li No applicable messages
and
aborts the command.
.Bl -tag -width delete
.It Ic \&\-
Print out the preceding message.
If given a numeric
argument
.Ar n  ,
goes to the
.Ar n Ns 'th
previous message and prints it.
.It Ic \&?
Prints a brief summary of commands.
.It Ic \&!
Executes the shell
(see
.Xr sh 1
and
.Xr csh 1 )
command which follows.
.It Ic Print
.Pq Ic P
Like
.Ic print
but also prints out ignored header fields.
See also
.Ic print ,
.Ic ignore
and
.Ic retain .
.It Ic Reply
.Pq Ic R
Reply to originator.
Does not reply to other
recipients of the original message.
.It Ic Type
.Pq Ic T
Identical to the
.Ic Print
command.
.It Ic alias
.Pq Ic a
With no arguments, prints out all currently-defined aliases.
With one
argument, prints out that alias.
With more than one argument, creates
a new alias or changes an old one.
.It Ic alternates
.Pq Ic alt
The
.Ic alternates
command is useful if you have accounts on several machines.
It can be used to inform
.Nm mail
that the listed addresses are really you.
When you
.Ic reply
to messages,
.Nm mail
will not send a copy of the message to any of the addresses
listed on the
.Ic alternates
list.
If the
.Ic alternates
command is given with no argument, the current set of alternate
names is displayed.
.It Ic chdir
.Pq Ic c
Changes the user's working directory to that specified, if given.
If
no directory is given, then changes to the user's login directory.
.It Ic copy
.Pq Ic co
The
.Ic copy
command does the same thing that
.Ic save
does, except that it does not mark the messages it
is used on for deletion when you quit.
.It Ic delete
.Pq Ic d
Takes a list of messages as argument and marks them all as deleted.
Deleted messages will not be saved in
.Ar mbox  ,
nor will they be available for most other commands.
.It Ic dp
(also
.Ic dt )
Deletes the current message and prints the next message.
If there is no next message,
.Nm mail
says
.Dq Li "at EOF" .
.It Ic edit
.Pq Ic e
Takes a list of messages and points the text editor at each one in
turn.
On return from the editor, the message is read back in.
.It Ic exit
.Pf ( Ic ex
or
.Ic x )
Effects an immediate return to the Shell without
modifying the user's system mailbox, his
.Ar mbox
file, or his edit file in
.Fl f  .
.It Ic file
.Pq Ic fi
The same as
.Ic folder  .
.It Ic folders
List the names of the folders in your folder directory.
.It Ic folder
.Pq Ic fo
The
.Ic folder
command switches to a new mail file or folder.
With no
arguments, it tells you which file you are currently reading.
If you give it an argument, it will write out changes (such
as deletions) you have made in the current file and read in
the new file.
Some special conventions are recognized for
the name.
# means the previous file, % means your system
mailbox, %user means user's system mailbox, & means
your
.Ar mbox
file, and
\&+\&folder means a file in your folder
directory.
.It Ic from
.Pq Ic f
Takes a list of messages and prints their message headers.
.It Ic headers
.Pq Ic h
Lists the current range of headers, which is an 18\-message group.
If
a
.Ql \&+
argument is given, then the next 18\-message group is printed, and if
a
.Ql \&\-
argument is given, the previous 18\-message group is printed.
.It Ic help
A synonym for
.Ic \&?
.It Ic hold
.Pf ( Ic ho ,
also
.Ic preserve )
Takes a message list and marks each
message therein to be saved in the
user's system mailbox instead of in
.Ar mbox  .
Does not override the
.Ic delete
command.
.It Ic ignore
Add the list of header fields named to the
.Ar ignored list .
Header fields in the ignore list are not printed
on your terminal when you print a message.
This
command is very handy for suppression of certain machine-generated
header fields.
The
.Ic Type
and
.Ic Print
commands can be used to print a message in its entirety, including
ignored fields.
If
.Ic ignore
is executed with no arguments, it lists the current set of
ignored fields.
.It Ic mail
.Pq Ic m
Takes as argument login names and distribution group names and sends
mail to those people.
.It Ic mbox
Indicate that a list of messages be sent to
.Ic mbox
in your home directory when you quit.
This is the default
action for messages if you do
.Em not
have the
.Ic hold
option set.
.It Ic next
.Pq Ic n
like
.Ic \&+
or
.Tn CR )
Goes to the next message in sequence and types it.
With an argument list, types the next matching message.
.It Ic preserve
.Pq Ic pre
A synonym for
.Ic hold  .
.It Ic print
.Pq Ic p
Takes a message list and types out each message on the user's terminal.
.It Ic quit
.Pq Ic q
Terminates the session, saving all undeleted, unsaved messages in
the user's
.Ar mbox
file in his login directory, preserving all messages marked with
.Ic hold
or
.Ic preserve
or never referenced
in his system mailbox, and removing all other messages from his system
mailbox.
If new mail has arrived during the session, the message
.Dq Li "You have new mail"
is given.
If given while editing a
mailbox file with the
.Fl f
flag, then the edit file is rewritten.
A return to the Shell is
effected, unless the rewrite of edit file fails, in which case the user
can escape with the
.Ic exit
command.
.It Ic reply
.Pq Ic r
Takes a message list and sends mail to the sender and all
recipients of the specified message.
The default message must not be deleted.
.It Ic respond
A synonym for
.Ic reply  .
.It Ic retain
Add the list of header fields named to the
.Ar retained list
Only the header fields in the retain list
are shown on your terminal when you print a message.
All other header fields are suppressed.
The
.Ic Type
and
.Ic Print
commands can be used to print a message in its entirety.
If
.Ic retain
is executed with no arguments, it lists the current set of
retained fields.
.It Ic save
.Pq Ic s
Takes a message list and a filename and appends each message in
turn to the end of the file.
The filename in quotes, followed by the line
count and character count is echoed on the user's terminal.
.It Ic set
.Pq Ic se
With no arguments, prints all variable values.
Otherwise, sets
option.
Arguments are of the form
.Ar option=value
(no space before or after =) or
.Ar option .
Quotation marks may be placed around any part of the assignment statement to
quote blanks or tabs, i.e.
.Dq Li "set indentprefix=\*q->\*q"
.It Ic saveignore
.Ic Saveignore
is to
.Ic save
what
.Ic ignore
is to
.Ic print
and
.Ic type  .
Header fields thus marked are filtered out when
saving a message by
.Ic save
or when automatically saving to
.Ar mbox  .
.It Ic saveretain
.Ic Saveretain
is to
.Ic save
what
.Ic retain
is to
.Ic print
and
.Ic type  .
Header fields thus marked are the only ones saved
with a message when saving by
.Ic save
or when automatically saving to
.Ar mbox  .
.Ic Saveretain
overrides
.Ic saveignore  .
.It Ic shell
.Pq Ic sh
Invokes an interactive version of the shell.
.It Ic size
Takes a message list and prints out the size in characters of each
message.
.It Ic source
The
.Ic source
command reads
commands from a file.
.It Ic top
Takes a message list and prints the top few lines of each.
The number of
lines printed is controlled by the variable
.Ic toplines
and defaults to five.
.It Ic type
.Pq Ic t
A synonym for
.Ic print  .
.It Ic unalias
Takes a list of names defined by
.Ic alias
commands and discards the remembered groups of users.
The group names
no longer have any significance.
.It Ic undelete
.Pq Ic u
Takes a message list and marks each message as
.Ic not
being deleted.
.It Ic unread
.Pq Ic U
Takes a message list and marks each message as
.Ic not
having been read.
.It Ic unset
Takes a list of option names and discards their remembered values;
the inverse of
.Ic set  .
.It Ic visual
.Pq Ic v
Takes a message list and invokes the display editor on each message.
.It Ic write
.Pq Ic w
Similar to
.Ic save  ,
except that
.Ic only
the message body
.Pq Ar without
the header) is saved.
Extremely useful for such tasks as sending and receiving source
program text over the message system.
.It Ic xit
.Pq Ic x
A synonym for
.Ic exit  .
.It Ic z
.Nm Mail
presents message headers in windowfuls as described under the
.Ic headers
command.
You can move
.Nm mail Ns 's
attention forward to the next window with the
.Ic \&z
command.
Also, you can move to the previous window by using
.Ic \&z\&\-  .
.El
.Ss Tilde/Escapes
.Pp
Here is a summary of the tilde escapes,
which are used when composing messages to perform
special functions.
Tilde escapes are only recognized at the beginning
of lines.
The name
.Dq Em tilde\ escape
is somewhat of a misnomer since the actual escape character can be set
by the option
.Ic escape .
.Bl -tag -width Ds
.It Ic \&~! Ns Ar command
Execute the indicated shell command, then return to the message.
.It Ic \&~b Ns Ar name ...
Add the given names to the list of carbon copy recipients but do not make
the names visible in the Cc: line ("blind" carbon copy).
.It Ic \&~c Ns Ar name ...
Add the given names to the list of carbon copy recipients.
.It Ic \&~d
Read the file
.Dq Pa dead.letter
from your home directory into the message.
.It Ic \&~e
Invoke the text editor on the message collected so far.
After the
editing session is finished, you may continue appending text to the
message.
.It Ic \&~f Ns Ar messages
Read the named messages into the message being sent.
If no messages are specified, read in the current message.
Message headers currently being ignored (by the
.Ic ignore
or
.Ic retain
command) are not included.
.It Ic \&~F Ns Ar messages
Identical to
.Ic \&~f ,
except all message headers are included.
.It Ic \&~h
Edit the message header fields by typing each one in turn and allowing
the user to append text to the end or modify the field by using the
current terminal erase and kill characters.
.It Ic \&~m Ns Ar messages
Read the named messages into the message being sent, indented by a
tab or by the value of
.Ar indentprefix  .
If no messages are specified,
read the current message.
Message headers currently being ignored (by the
.Ic ignore
or
.Ic retain
command) are not included.
.It Ic \&~M Ns Ar messages
Identical to
.Ic \&~m ,
except all message headers are included.
.It Ic \&~p
Print out the message collected so far, prefaced by the message header
fields.
.It Ic \&~q
Abort the message being sent, copying the message to
.Dq Pa dead.letter
in your home directory if
.Ic save
is set.
.It Ic \&~r Ns Ar filename
Read the named file into the message.
.It Ic \&~s Ns Ar string
Cause the named string to become the current subject field.
.It Ic \&~\&t Ns Ar name ...
Add the given names to the direct recipient list.
.It Ic \&~\&v
Invoke an alternate editor (defined by the
.Ev VISUAL
option) on the
message collected so far.
Usually, the alternate editor will be a
screen editor.
After you quit the editor, you may resume appending
text to the end of your message.
.It Ic \&~w Ns Ar filename
Write the message onto the named file.
.It Ic \&~\&| Ns Ar command
Pipe the message through the command as a filter.
If the command gives
no output or terminates abnormally, retain the original text of the
message.
The command
.Xr fmt 1
is often used as
.Ic command
to rejustify the message.
.It Ic \&~: Ns Ar mail-command
Execute the given mail command.
Not all commands, however, are allowed.
.It Ic \&~~ Ns Ar string
Insert the string of text in the message prefaced by a single ~.
If
you have changed the escape character, then you should double
that character in order to send it.
.El
.Ss Mail Options
Options are controlled via
.Ic set
and
.Ic unset
commands.
Options may be either binary, in which case it is only
significant to see whether they are set or not; or string, in which
case the actual value is of interest.
The binary options include the following:
.Bl -tag -width append
.It Ar append
Causes messages saved in
.Ar mbox
to be appended to the end rather than prepended.
This should always be set (perhaps in
.Pa /etc/mail.rc ) .
.It Ar ask, asksub
Causes
.Nm mail
to prompt you for the subject of each message you send.
If
you respond with simply a newline, no subject field will be sent.
.It Ar askcc
Causes you to be prompted for additional carbon copy recipients at the
end of each message.
Responding with a newline indicates your
satisfaction with the current list.
.It Ar askbcc
Causes you to be prompted for additional blind carbon copy recipients at the
end of each message.
Responding with a newline indicates your
satisfaction with the current list.
.It Ar autoprint
Causes the
.Ic delete
command to behave like
.Ic dp
\- thus, after deleting a message, the next one will be typed
automatically.
.It Ar debug
Setting the binary option
.Ar debug
is the same as specifying
.Fl d
on the command line and causes
.Nm mail
to output all sorts of information useful for debugging
.Nm mail  .
.It Ar dot
The binary option
.Ar dot
causes
.Nm mail
to interpret a period alone on a line as the terminator
of a message you are sending.
.It Ar hold
This option is used to hold messages in the system mailbox
by default.
.It Ar ignore
Causes interrupt signals from your terminal to be ignored and echoed as
@'s.
.It Ar ignoreeof
An option related to
.Ar dot
is
.Ar ignoreeof
which makes
.Nm mail
refuse to accept a control-d as the end of a message.
.Ar Ignoreeof
also applies to
.Nm mail
command mode.
.It Ar metoo
Usually, when a group is expanded that contains the sender, the sender
is removed from the expansion.
Setting this option causes the sender
to be included in the group.
.It Ar noheader
Setting the option
.Ar noheader
is the same as giving the
.Fl N
flag on the command line.
.It Ar nosave
Normally, when you abort a message with two
.Tn RUBOUT
(erase or delete)
.Nm mail
copies the partial letter to the file
.Dq Pa dead.letter
in your home directory.
Setting the binary option
.Ar nosave
prevents this.
.It Ar Replyall
Reverses the sense of
.Ic reply
and
.Ic Reply
commands.
.It Ar quiet
Suppresses the printing of the version when first invoked.
.It Ar searchheaders
If this option is set, then a message-list specifier in the form ``/x:y''
will expand to all messages containing the substring ``y'' in the header
field ``x''.  The string search is case insensitive.
.It Ar verbose
Setting the option
.Ar verbose
is the same as using the
.Fl v
flag on the command line.
When mail runs in verbose mode,
the actual delivery of messages is displayed on the user's
terminal.
.El
.Ss Option String Values
.Bl -tag -width Va
.It Ev EDITOR
Pathname of the text editor to use in the
.Ic edit
command and
.Ic \&~e
escape.
If not defined, then a default editor is used.
.It Ev LISTER
Pathname of the directory lister to use in the
.Ic folders
command.
Default is
.Pa /bin/ls .
.It Ev PAGER
Pathname of the program to use in the

NEWS
S - n a i l / S - m a i l x  N e w s
====================================

  mdocmx(7) anchors are denoted by a number-sign #:
  pointing a browser to the "#ANCHOR" of the online manual works.

v14.10.0-alpha ("Mountains O' Things"), 2026-04-08
--------------------------------------------------

Credits, in order of commit appearance:
Maxim Tarasov, Geoff Clare, Gavin Troy, Olaf Hering,
Stuart Henderson, Olav Mørkrid, John Levine, Z, Vincent Lefevre,
Oskari Pirhonen, Robert Elz, Florian Weimer, Adhemerval Zanella,
Damien Miller, Madou Mad, Giannis-2, Qualys (Security Advisory
team), Ross Burton, Mingli Yu, Hans Bezemer, Ashton Fagg,
Steve Blinkhorn, Arne Wörner, Harald van Dijk, Stephen Isard,
Jens Schleusener, John Holder, Χάρης Καραχριστιανίδης (Hariskar),
Denys Vlasenko, Dr. Werner Fink, Predrag Punosevac,
Sergey Matveev, Solar Designer, Kevin Bloom, Arkadiusz Miśkiewicz,
Thorsten Glaser, Claus Assmann, Robin Humble, Chet Ramey,
Tomáš Korbař, Viktor Dukhovni, Martin Uecker, Dag-Erling Smørgrav,
Kyle Evans, rahl, Mia Soweli, and Ryan Matthews.

We welcome Maxim Tarasov, Olaf Hering, John Levine, Z,
Oskari Pirhonen, Florian Weimer, Adhemerval Zanella, Damien Miller,
Giannis-2, Qualys (Security Advisory team), Ross Burton, Mingli Yu,
Hans Bezemer, Ashton Fagg, Steve Blinkhorn, Arne Wörner,
Denys Vlasenko, Kevin Bloom, Arkadiusz Miśkiewicz, Thorsten Glaser,
Robin Humble, Tomáš Korbař, Viktor Dukhovni, Martin Uecker,
Dag-Erling Smørgrav, rahl, Mia Soweli, and Ryan Matthews in THANKS.

We are https://scan.coverity.com/projects/s-nail (project 444).

PACKAGER NOTES (make.rc)
^^^^^^^^^^^^^^^^^^^^^^^^

- New OPT_NET_TEST, builds a test helper for network protocol
  functionality, default enabled (with OPT_NET=y).

- Whoever uses OPT_AUTOCC (default) may now use VAL_AUTOCC to
  fine-tune settings.  The "option_care" flag may be of special
  interest, the default though is
    VAL_AUTOCC=bind_now,pie,stackprot,fortify,cfprot
  (Stuart Henderson)

- OPT_DOCSTRINGS and OPT_UISTRINGS were removed.

- OPT_OAUTH_HELPER, external OAuth Python3 helper program, default
  enabled (with OPT_NET=y)

- OPT_BUILD_TESTS umbrella option that covers all possible
  programs which are build solely as helper for the test.

- VAL_BUILD_INFO may be set by packagers who do not want to
  include build information in the binary.  (Mingli Yu)
==alpha

NOTES, ChangeLog
^^^^^^^^^^^^^^^^

+ In general: if it did not work in the past, try it now.
  (There are way too many bugfixes to be noted.)

+ v15-compat=y is now default!  (Setting will disappear in v15.)

+ Command line interface moved towards sh(1)ell-style quite a bit.

  This includes that `if'#226 .. series now practically works
  identical to bash(1)s [[ ]] construct, such things now succeed:
    uns i; if -z $i; ec ya; end

+ Several settings gained built-in initial values.  Like this non-
  expert users should get away out-of-the-box more easily.
  (Arkadiusz Miśkiewicz, Kevin Bloom)
  Note: includes *mime-force-sendout*#498 (Arkadiusz Miśkiewicz)

  ! Remarks: power users may not like certain of these, for example,
  ! that *emptystart*#432 is now set by default.  So far there is no
  ! simple toggle to turn these extensions all off, like the new
  ! *expert*#439 that Norman Ramsey asked for.
  ! I wonder whether that should change with the final v14.10.0
  ! release.  For at least -alpha however they have to be
  ! disabled/changed explicitly, in resource files, for example.

  (The entire set of initial settings is shown by eg
    s-nail -:/ -Xse -Xx
  The entire set of known variables plus initial settings:
    s-nail -:/ [-v] -Xvars -Xx)

+ `xcall'#317, the "manual tail call optimization", no longer
  causes memory lingering in the xcall machinery (and more fixes).
  Oh no, we can implement yes(1)!
    s-nail -:/ -X'define yes {
      ec yes
      xcall yes
    }
    call yes'

+ SMTP configuration changed completely via new *smtp-config*#590,
  including support of SMTP extensions like PIPELINING and more.
  (It is likely "it just works" for you out of the box.  Any hint
  if not would be appreciated!)

  ! POP3 and IMAP are not yet converted to the new syntax in -alpha.

+ The old Heirloom-style approach of credentials was removed.
  It was obsoleted since before 2015.

+ SMTP, POP3 and IMAP gained support for OAUTHBEARER
  authentication mechanisms.
  (IMAP also PLAIN.  (Stephen Isard))

  There is an optional *build-oauth-helper*#407 Python3 script
  that can be used by the macros shown in the new XY-config(7)
  manual, which gives copy+paste examples, to manage OAuth
  credentials.

  ! With -alpha, credentials must be kept in a "hot" and valid
  ! state; the release hopefully will call out for credentials
  ! when they are *really* needed.

+ MIME classification learned about chains like .tar.gz.
  New ?o [only-handler] `mimetype'#235 marker, a good thing.
  Both: (Olaf Hering)
  Also new ?s [send-text] marker.

    Because MIME media type handlers, like those defined in "The
    Mailcap files", match media types, non-standardized fictional
    types like "x-tar-gz" are used in the wild as MIME environments
    become configured.

    That is, IANA MIME registry standards do not know about "extension
    chains": "tar.gz" is thus a gzip(1) compressed file, and the sane
    solution of recursively unpacking until no more MIME media type
    unpacking is possible is not available.

    Therefore introduce this type-marker, which says that the given
    MIME media type shall only be matched when looking for handlers,
    but not when classifying content to create messages.

  Some are built-in by default, see
    $ s-nail -:/ -Xmimet -Xx | grep only-hand

+ File related functions of `vexpr'#313 moved to new `fop'#213.
  Many `csop'#172, `fop' and `vexpr' functions added or extended.

+ New $(( )) shell-style arithmetic evaluator.
  ! Known -alpha bug for tremendously deep ?: style conditionals:
  !  ok: $((0?2:(3,0)?(4,5):(5,0)?(7,8):(9,10)))

usr/src/s2/mail.c
#
/* mail command usage
	mail
		prints your mail
	mail people
		sends standard input to people
 */

#define	SIGINT	2

struct	utmp {
	char	name[8];
	char	tty;
	char	pad1;
	int	ltime[2];
	int	pad2;
};

struct	passwd {
	char	*pw_name;
	char	*pw_passwd;
	int	pw_uid;
	int	pw_gid;
	char	*pw_gecos;
	char	*pw_dir;
	char	*pw_shell;
};

char	lettmp[] "/tmp/maxxxxx";
char	preptmp[] "/tmp/mbxxxxx";
int	pwfil;

main(argc, argv)
char **argv;
{
	register me;
	extern fout;
	register struct passwd *p;
	register char *cp;
	static struct utmp ubuf;
	int uf;

	maketemp();
	if (argc==1 || argc==2 && argv[1][0]=='-') {
		printmail(argc, argv);
		delexit();
	}
	signal(SIGINT, delexit);
	fout = creat(lettmp, 0600);
	if (((me=ttyn(1))!='x' || (me=ttyn(2))!='x')
	 && (uf = open("/etc/utmp", 0)) > 0) {
		while (read(uf, &ubuf, sizeof ubuf) == sizeof ubuf)
			if (ubuf.tty == me) {
				ubuf.name[8] = ' ';
				close(uf);
				for (cp=ubuf.name; *cp++!=' ';);
				*--cp = 0;
				bulkmail(argc, argv, ubuf.name);
			}
	}
	me = getuid() & 0377;
	setpw();
	for (;;)
		if ((p = getpwent()) && p->pw_uid == me)
			bulkmail(argc, argv, p->pw_name);
	fout = 1;
	printf("Who are you?\n");
	delexit();
}

printmail(argc, argv)
char **argv;
{
	extern fin, fout;
	register n, c, f;

	if (fopen(".mail", &fin)>=0 && (c = getchar())) {
		do {
			putchar(c);
		} while (c = getchar());
		close(fin);
		c = 'y';
		if (argc<2) {
			if (ttyn(0)!='x') {
				printf("Save?");
				fin = 0;
				c = getchar();
			}
		} else
			c = argv[1][1];
		if (c=='y') {
			prepend(".mail", "mbox");
			printf("Saved mail in 'mbox'\n");
		}
		close(creat(".mail"));
	} else
		printf("No mail.\n");
}

bulkmail(argc, argv, from)
char **argv, *from;
{
	extern fin, fout;
	int tbuf[2];
	register c;

	fin = 0;
	(&fin)[1] = 0;
	time(tbuf);
	printf("From %s %s", from, ctime(tbuf));
	while (c = getchar())
		putchar(c);
	putchar('\n');
	flush();
	close(fout);
	while (--argc > 0)
		sendto(*++argv);
	delexit();
}

sendto(person)
char *person;
{
	static saved;
	extern fout;
	extern fin;
	register struct passwd *p;

	setpw();
	while (p = getpwent()) {
		if (equal(p->pw_name, person)) {
			if (prepend(lettmp, cat(p->pw_dir, "/.mail"))==0)
				break;
			return;
		}
	}
	fout = 1;
	flush();
	printf("Can't send to %s.\n", person);
	if (ttyn(0)!='x' && saved==0) {
		unlink("dead.letter");
		saved++;
		printf("Letter saved in 'dead.letter'\n");
		prepend(lettmp, "dead.letter");
	}
}

prepend(from, to)
char *from, *to;
{
	extern int fin, fout;

	fcreat(preptmp, &fout);
	fopen(from, &fin);
	while (putchar(getchar()));
	close(fin);
	fopen(to, &fin);
	while (putchar(getchar()));
	close(fin);
	flush();
	close(fout);
	if (fcreat(to, &fout) < 0) {
		fout = 1;
		return(0);
	}
	fopen(preptmp, &fin);
	while(putchar(getchar()));
	flush();
	close(fout);
	close(fin);
	fout = 1;
	return(1);
}

setpw()
{
	extern fin;

	if (pwfil == 0) {
		fopen("/etc/passwd", &fin);
		pwfil = fin;
	} else
		fin = pwfil;
	(&fin)[1] = 0;
	seek(fin, 0, 0);
}

getpwent()
{
	register char *p;
	register c;
	static struct passwd passwd;
	static char line[100];
	extern fin;

	p = line;
	while((c=getchar()) != '\n') {
		if(c <= 0)
			return(0);
		if(p < line+98)
			*p++ = c;
	}
	*p = 0;
	p = line;
	passwd.pw_name = p;
	p = pwskip(p);
	passwd.pw_passwd = p;
	p = pwskip(p);
	passwd.pw_uid = atoi(p);
	p = pwskip(p);
	passwd.pw_gid = atoi(p);
	p = pwskip(p);
	passwd.pw_gecos = p;
	p = pwskip(p);
	passwd.pw_dir = p;
	p = pwskip(p);
	passwd.pw_shell = p;
	return(&passwd);
}

pwskip(ap)
char *ap;
{
	register char *p;

	p = ap;
	while(*p != ':') {
		if(*p == 0)
			return(p);
		p++;
	}
	*p++ = 0;
	return(p);
}

delexit()
{
	unlink(lettmp);
	unlink(preptmp);
	exit(0);
}

maketemp()
{
	int i, pid, d;

	pid = getpid();
	for (i=11; i>=7; --i) {
		d = (pid&07) + '0';
		lettmp[i] = d;
		preptmp[i] = d;
		pid =>> 3;
	}
}

equal(as1, as2)
{
	register char *s1, *s2;

	s1 = as1;
	s2 = as2;
	while (*s1++ == *s2)
		if (*s2++ == 0)
			return(1);
	return(0);
}

cat(ap1, ap2)
char *ap1, *ap2;
{
	register char *p1, *p2;
	static char fn[32];

	p1 = ap1;
	p2 = fn;
	while (*p2++ = *p1++);
	p2--;
	p1 = ap2;
	while (*p2++ = *p1++);
	return(fn);
}

